tfs/writer.py

```diff
@@ -147,85 +167,137 @@ def _insert_index_column(data_frame: TfsDataFrame | pd.DataFrame, save_index: st
+            (int, float, str, potentially bool, complex etc.) and the type
+            of this value is used to infer the formatting.
+        width (int): column width to use when formatting the header line.
+
+    Returns:
+        The full, formatted header line string.
+    """
     if not isinstance(name, str):
         errmsg = f"{name} is not a string"
         raise TypeError(errmsg)
     type_str = _value_to_type_string(value)
-    if type_str == "%s":
-        value = f'"{value}"'
-    return f"@ {name:<{width}} {type_str} {value:>{width}}"
+    value_str = TfsStringFormatter().format_field(value, _value_to_string_format_id(value))
+    return f"@ {name:<{width}} {type_str} {value_str:>{width}}"
 
 
-def _get_colnames_string(
-    colnames: list[str], colwidth: int, left_align_first_column: bool  # noqa: FBT001, FBT002
-) -> str:
+def _get_colnames_string(colnames: list[str], colwidth: int, left_align_first_column: bool) -> str:  # noqa: FBT001
+    """Returns the string for the line with the column names."""
     format_string = _get_row_format_string([None] * len(colnames), colwidth, left_align_first_column)
     return "* " + format_string.format(*colnames)
 
 
-def _get_coltypes_string(
-    types: pd.Series, colwidth: int, left_align_first_column: bool  # noqa: FBT001, FBT002
-) -> str:
+def _get_coltypes_string(types: pd.Series, colwidth: int, left_align_first_column: bool) -> str:  # noqa: FBT001
+    """Returns the string for the line with the column type specifiers."""
     fmt = _get_row_format_string([str] * len(types), colwidth, left_align_first_column)
-    return "$ " + fmt.format(*[_dtype_to_id_string(type_) for type_ in types])
+    return "$ " + fmt.format(*[_dtype_to_tfs_format_id(type_) for type_ in types])
 
 
 def _get_data_string(
     data_frame: TfsDataFrame | pd.DataFrame,
     colwidth: int,
     left_align_first_column: bool,  # noqa: FBT001
 ) -> str:
+    """
+    Returns the complete string to be written for the data part of the dataframe.
+    This corresponds to all the data rows, after the column names and the column
+    type specifiers.
+
+    Args:
+        data_frame (TfsDataFrame | pd.DataFrame): the dataframe to write.
+        colwidth (int): column width to use when formatting the data.
+        left_align_first_column (bool): whether to left-align the first column or not.
+
+    Returns:
+        The full string representation of the data part of the dataframe.
+    """
     if len(data_frame.index) == 0 or len(data_frame.columns) == 0:
         return "\n"
     format_strings = "  " + _get_row_format_string(data_frame.dtypes, colwidth, left_align_first_column)
-    data_frame = _quote_string_columns(data_frame)
     data_frame = data_frame.astype(object)  # overrides pandas auto-conversion (lead to format bug)
-    return "\n".join(data_frame.apply(lambda series: format_strings.format(*series), axis=1))
+    string_formatter = TfsStringFormatter()
```

Comment:
**FEATURE**

Same thing here, we let the formatter class handle everything for us.